### PR TITLE
Delete stale non-e2e devices for users, take 3

### PR DESCRIPTION
**Pull request by erikjohnston**
This should help reduce the number of devices e.g. simple bots the repeatedly login rack up.

We only delete non-e2e devices as they should be safe to delete, whereas if we delete e2e devices for a user we may accidentally break their ability to receive e2e keys for a message.

Basically #14595, but makes sure that we only have one delete going per-person at a time, and with a longer timeout.

### CODE REVIEW

File: synapse/handlers/register.py [hunk 1]
```diff
@@ -110,6 +112,10 @@ def __init__(self, hs: "HomeServer"):
         self._server_notices_mxid = hs.config.servernotices.server_notices_mxid
         self._server_name = hs.hostname
 
+        # The set of users that we're currently pruning devices for. Ensures
+        # that we don't have two such jobs running at once.
+        self._currently_pruning_devices_for_users: Set[str] = set()
```
Comment (clokep) [inline synapse/handlers/register.py:117]:
```suggestion
        # The set of users that we're currently pruning devices for. Ensures
        # that we don't have two such jobs for the same user running at once.
        self._currently_pruning_devices_for_users: Set[str] = set()
```

I wonder if we should cap the number of users we allow at once? Or have a single queue we add them to and chunk through them in a background task, so we would only process a single user at once?

File: synapse/handlers/register.py [hunk 2]
```diff
@@ -845,6 +854,9 @@ class and RegisterDeviceReplicationServlet.
         # This can only run on the main process.
         assert isinstance(self.device_handler, DeviceHandler)
 
+        # Prune the user's device list if they already have a lot of devices.
+        await self._prune_too_many_devices(user_id)
```
Comment (clokep) [inline synapse/handlers/register.py:858]: Maybe this is clearer?

```suggestion
        # Prune the user's device list if they already have a lot of devices.
        await self._maybe_prune_too_many_devices(user_id)
```

File: synapse/storage/databases/main/devices.py [hunk 3]
```diff
@@ -1602,6 +1602,72 @@ def _txn(txn: LoggingTransaction) -> int:
...
+            keyvalues={"user_id": user_id, "hidden": False},
+            retcol="COALESCE(COUNT(*), 0)",
+            desc="count_devices",
+        )
+
+        # We let users have up to ten devices without pruning.
+        if num_devices <= 10:
+            return []
+
+        # We prune everything older than N days.
```
Comment (clokep) [inline synapse/storage/databases/main/devices.py:1623]:
```suggestion
        # We prune everything older than 14 days.
```

Although this isn't really true based on the logic below with the 50 devices?

File: synapse/storage/databases/main/devices.py [hunk 4]
```diff
@@ -1602,6 +1602,72 @@ def _txn(txn: LoggingTransaction) -> int:
...
+
+        num_devices = await self.db_pool.simple_select_one_onecol(
+            table="devices",
+            keyvalues={"user_id": user_id, "hidden": False},
+            retcol="COALESCE(COUNT(*), 0)",
+            desc="count_devices",
+        )
+
+        # We let users have up to ten devices without pruning.
+        if num_devices <= 10:
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py:1620]: I've updated the comments, but basically: we always delete devices not seen in the last 14 days, but we *also* cap the maximum number of devices to 50.

File: synapse/handlers/register.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/register.py]: Yeah, good point. I've added a basic limiter to cap it at 5 users at a time :shrug:

File: synapse/storage/databases/main/devices.py [hunk 5]
```diff
@@ -1602,6 +1602,73 @@ def _txn(txn: LoggingTransaction) -> int:
...
+            return []
+
+        # We always prune devices not seen in the last 14 days...
+        max_last_seen = self._clock.time_msec() - 14 * 24 * 60 * 60 * 1000
+
+        # ... but we also cap the maximum number of devices the user can have to
+        # 50.
+        if num_devices > 50:
+            # Chose a last seen that ensures we keep at most 50 devices.
```
Comment (clokep) [inline synapse/storage/databases/main/devices.py:1629]:
```suggestion
            # Choose a last seen that ensures we keep at most 50 devices.
```

Comment (clokep) [reply synapse/storage/databases/main/devices.py:1629]: I think this is also "at most 50 non-hidden, non-E2E devices"?

**Review by clokep — APPROVED** [review]
Lets give it a go.

**Comment by erikjohnston** [comment]
Waiting to merge this until we are ready to put this on matrix.org, as previous iterations of this change have caused production issues.

**Pull request merged**